internal/controller/driver_controller.go

```diff
@@ -18,45 +18,1033 @@ package controller
+	reconcileHandler := driverReconcile{}
+	reconcileHandler.DriverReconciler = *r
+	reconcileHandler.ctx = ctx
+	reconcileHandler.log = ctrllog.FromContext(ctx)
+	reconcileHandler.driver.Name = req.Name
+	reconcileHandler.driver.Namespace = req.Namespace
+
+	return reconcileHandler.reconcile()
+}
+
+func (r *driverReconcile) reconcile() (ctrl.Result, error) {
+	r.log.Info("Enter Reconcile", "req", client.ObjectKeyFromObject(&r.driver))
+
+	// Load the driver desired state based on driver resource, operator config resource and default values.
+	if err := r.LoadAndValidateDesiredState(); err != nil {
+		return ctrl.Result{}, err
+	}
 
-	// TODO(user): your logic here
+	// Concurrently reconcile different aspects of the clusters actual state to meet
+	// the desired state defined on the driver object
+	errChan := utils.RunConcurrently(
+		r.reconcileK8sCsiDriver,
+		r.reconcileControllerPluginDeployment,
+		r.reconcileNodePluginDeamonSet,
+		r.reconcileLivnessService,
+	)
+
+	// Check if any reconcilatin error where raised during the concurrent execution
+	// of the reconciliation steps.
+	errList := utils.ChannelToSlice(errChan)
+	if err := errors.Join(errList...); err != nil {
+		r.log.Error(err, "Reconciliation failed")
+		return ctrl.Result{}, err
+	}
 
 	return ctrl.Result{}, nil
 }
 
-// SetupWithManager sets up the controller with the Manager.
-func (r *DriverReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&csiv1alpha1.Driver{}).
-		Complete(r)
+func (r *driverReconcile) LoadAndValidateDesiredState() error {
+	// Validate that the requested name for the CSI driver isn't already claimed by an existing CSI driver
+	// (Can happen if a driver with an identical name was created in a different namespace)
+	if err := r.Get(r.ctx, client.ObjectKey{Name: r.driver.Name}, &storagev1.CSIDriver{}); err == nil {
+		r.log.Error(err, "Desired name already in use by a different CSI Driver", "name", r.driver.Name)
+		return nil
+	} else if !k8sErrors.IsNotFound(err) {
+		r.log.Error(err, "Failed to query the existence of a CSI Driver", "name", r.driver.Name)
+		return err
+	}
+
+	// Load operator configuration resource
+	opConfig := csiv1a1.OperatorConfig{}
+	opConfig.Name = operatorConfigName
+	opConfig.Namespace = operatorNamespace
+	if err := r.Get(r.ctx, client.ObjectKeyFromObject(&opConfig), &opConfig); client.IgnoreNotFound(err) != nil {
+		r.log.Error(err, "Unable to load operatorconfig.csi.ceph.io", "name", client.ObjectKeyFromObject(&opConfig))
+		return err
+	}
+
+	// Extract the driver sort name and driver type
+	matches := nameRegExp.FindStringSubmatch(r.driver.Name)
+	if len(matches) != 3 {
+		return fmt.Errorf("invalid driver name")
+	}
+	r.driverPrefix = matches[1]
+	r.driverType = DriverType(strings.ToLower(matches[2]))
+
+	// Load the current desired state in the form of a ceph csi driver resource
+	if err := r.Get(r.ctx, client.ObjectKeyFromObject(&r.driver), &r.driver); err != nil {
```

Comment:
maybe better to have helpers from the start `r.{get,list,update,delete}`?

Reply:
It has not proven very useful over the last 2-3 projects where we implemented them. It just created confusion for the reader and another layer of indirection. On many commits I saw contributors reverting to using `r.Get` instead of `r.get`.

Also, please do not assume that we will keep the context the same on all operations, mainly as I do want to introduce concurrency controls later on.

Reply:
resolve.